### DISCUSSION TITLE
Assigning enums to integral types

### DISCUSSION THREAD

**Discussion by Pyrdacor** [dotnet/csharplang#6488] [Q&A] [answered]
Why is an explicit cast needed when assigning an enum value to a variable of its underlying integral type?

```cs
enum Color { Red, Green }
enum Animal : byte { Fish, Ape }

int color = Color.Red; // CS0266
byte animal = Animal.Ape; // CS0266
```

This feels like a weird unnecessary limitation which is most likely caused by old design choices. But wouldn't it be possible to just allow this? I mean the compiler could just add the explicit cast under the hood itself.

The whole working with enum values is much too complicated imo. For example flag enums are often used in binary data formats and therefore are often serialized as integral types like bytes. It is really a pain in the ... to add all those explicit casts even though the compiler totally knows the integral types at play. I mean the compiler can do all the checks for integral types but fails for enums which are basically named integral values.

Did I miss something? Is it planned to improved something in the field of enums in the near future?

**Comment by HaloFour** [comment, marked as answer]
Enums aren't supposed to be treated as raw values of the underlying type and I don't think it's worth making it easier to accidentally to do in order to remove an explicit cast from the uncommon use cases such as binary serialization.

**Reply by Pyrdacor** [reply]
> Enums aren't supposed to be treated as raw values of the underlying type

This is not what I am aiming for. But it is obvious that the value range of an enum is at max the value range of the underlying type. Every possible value of an enum with underlying type int, is represented by the value of an int. That is basically what the underlying type expresses. So it should be totally valid to assign those values directly to an int. I can also assign a byte to an int but a byte is not treated as an int. The value range of int just includes all bytes. The same is true for the enum.

**Reply by CyrusNajmabadi** [reply]
The reason is so that the conversiom is done explicitly with thought from the user that this is the right thing to do. This is part of what enums are. They are but intended for going back and forth implicitly.

**Comment by Pyrdacor** [comment]
What do you mean by "accidentally"? Do you have an example where you could do this accidentally in a way that is somewhat harmful or leads to logical errors?

The binary serialization was just an example. It isn't by far the only place where you would assign an integral value to an enum or vice versa. In the current state of possibilities I even am forced to store enum values in some integral type when I have to perform some advanced logic on them.

Is the purpose of the underlying type solely the limitation of the possible values?

**Reply by CyrusNajmabadi** [reply]
```c#
var size = ...
var list = new List(capacity: size);
```

Except here 'size' is an enum, not an integer. And implicitly converting here would lead to a bug (esp if the enum has negative values).  Forcing a cast makes me think about what I'm doing here.

**Reply by Pyrdacor** [reply]
This is a very unlikely scenario. Why would you pass an enum to a list constructor?

A much more common use case is that you are dealing with enum values and manipulate them.

There are times where you have to perform arithmetic logic on enum values. Or you have to pass an int to some interface (of some API) but want to work with enums from you own codebase etc.

```cs
enum Color
{
    Black,
    Red,
    Green,
    Blue,
    FirstUserColor 
}

int colorIndex = controlType switch
{
    Button => (int)Color.Red,
    // ... 100 more casts or a cast of the whole expression, but some cases return ints like the default one
    UserControl => (int)(Color.FirstUserColor + control.ColorIndex),
    _ => 0
};

someFramework.RenderControl(colorIndex); // method requires an int
```

**Reply by Pyrdacor** [reply]
At least initializations should be safe without the cast as you know what you are dealing with. Usage of the enum value somewhere else might be somewhat confusing but an initialization of an int to an enum?


Here another example. I have to write many console apps (tools) with a bunch of specific error codes. I tend to use enums but the return type of the main function is int as it can't know my enum. This is often the case for interfaces which except more generic types or values but you want to be specific in your own project.

Enums help to make values more expressive than plain numbers.

```cs
enum Error
{
    NoError,
    AccessDenied,
    ...
}

int errorCode = Error.NoError;

// Some logic

Environment.Exit(errorCode);
```

**Reply by HaloFour** [reply]
It's an intentional part of the design of enums to not treat them as the underlying integral value, and disallowing implicit casts is part of that design.  Wanting to perform arithmetic operations on an enum is a niche case, one that shouldn't be encouraged by making it easier.

**Reply by Pyrdacor** [reply]
Ok as this is by design, I will have to accept this. Thanks for clarification.

**Reply by CyrusNajmabadi** [reply]
> This is a very unlikely scenario. Why would you pass an enum to a list constructor?

It could happen to any constructor or method accepting an integer. And the reason I did it is because I confused an enum representing scale with an integer representing something else. Without the error I could easily make this mistake. With the error I have to stop and think and only cast if I'm sure it is sensible. 

I would ask the opposite. It seems non sensical to treat enums as their underlying values (including for serialization where I have to think about endianness). So if want a super compelling reason to get rid of this useful error.

**Reply by Pyrdacor** [reply]
Maybe we live in different worlds regarding use cases. I develop games and tools around their game data. Here enums represent values that almost always match bytes inside the game data files. In the majority of cases normal enums are bytes and flagged enums are words or double words. I do this for over 20 years and others too. C and C++ do this too. So here enums are just more expressive versions of a byte, word or dword. They are a tool to work with those underlying types in a way which is better readable. Serialization might involve additional logic. In my case I often use BE. But even there I have to cast a lot. I never ever had a single case where I confused an enum with an int where it would make problems. So maybe I am not the "general user" of C#. I am more in the low level corner and still likes aspects of C and am happy that C# allows the low level stuff and adding cool new features around it.

**Reply by Pyrdacor** [reply]
I also work as a full stack dev but as the data comes solely from tables, I've never seen a single enum in my daily work.

The other things I can think of using enums are UIs. But even there you will have some low level interface where you will have to pass your own enums as ints or other enums.

**Reply by CyrusNajmabadi** [reply]
> They are a tool to work with those underlying types in a way which is better readable

I don't really see then how that's any better than just naming a constant underlying type.  What is the enum getting you?  For me, the enum *is* getting me the separated domains.

**Reply by CyrusNajmabadi** [reply]
>  I am more in the low level corner and still likes aspects of C

This seems like a space where you'd want even clearer separation.  The low-level stuff is often extremely permissive, allowing almost anything.  Having no guardrails means trivially passing in enums where they would be non-sensical.  For example, any of the low level APIs that take ints/nints/whatever really should not be passed something like ConsoleColor.XXX.  It's just non-sensical.  Stopping it seems desirable.  

> In the majority of cases normal enums are bytes 

I don't see how that makes it better.  Say i have game enums X and Y.  Keeping those domains separated and not accidentally just losing their strong typing and becoming ints where they might intermix inappropriately, seems value to enforce.

**Reply by Pyrdacor** [reply]
For me enums define valid subsets of another type. For example only some integers. It is nearly like some compile-time constant dictionary where the enum value name is the key and the value is the value. In a perfect world I would even like to have enums for strings or even objects. But this is another story.

Basically you define a range of valid values and give those values names. The benefit over just constants is that you have a real type to express this subset.

It is a type which represents compile-time constant values of a given value range.

Here some fictional example

```
enum DataFile : string
{
    Graphics = "Gfx.zip",
    Music = "Music.zip"
}

Stream ReadFile(DataFile dataFile)
{
    // ...
}
```

I think this is a convenient system. This could even be used with objects etc. And it is the same for me with integers. I can define a subset in the domain of an integer to be valid and create a type for this expression. Still it is an integer as all values are integers. So I also want to be able to treat them as integers.

**Reply by CyrusNajmabadi** [reply]
> Still it is an integer as all values are integers. So I also want to be able to treat them as integers.

I'm curious why though.  What's the value in treating it like an integer?  Can it be a length?  Can it be added to other values in that enum?  What does it mean to divide two of the enum values together?

**Reply by Pyrdacor** [reply]
> > I am more in the low level corner and still likes aspects of C
> 
> This seems like a space where you'd want even clearer separation. The low-level stuff is often extremely permissive, allowing almost anything. Having no guardrails means trivially passing in enums where they would be non-sensical. For example, any of the low level APIs that take ints/nints/whatever really should not be passed something like ConsoleColor.XXX. It's just non-sensical. Stopping it seems desirable.
> 

I agree. You have a point there. My intention is more about allowing the concrete enum values as integers and not so much variables of that enum.

So I think about `CallSomeFunction(ConsoleColor.Red)` and not so much `CallSomeFunction(myVar)`.

Things like `int color = ConsoleColor.Red;` should be allowed. They are not really prone to errors and the compiler could just treat them as integers here. Here the intention is very obvious as the int type is already specified explicit on the left.

> > In the majority of cases normal enums are bytes
> 
> I don't see how that makes it better. Say i have game enums X and Y. Keeping those domains separated and not accidentally just losing their strong typing and becoming ints where they might intermix inappropriately, seems value to enforce.

I am not sure if I get you right. I don't want everything to become ints. The enums stay as they are and can't be converted into other enums. I just want to ease their use when I have to store them or treat them as the underlying type. For example when I have to add an offset to them, increment them or assign them to the byte they are written to.

**Reply by Pyrdacor** [reply]
> > Still it is an integer as all values are integers. So I also want to be able to treat them as integers.
> 
> I'm curious why though. What's the value in treating it like an integer? Can it be a length? Can it be added to other values in that enum? What does it mean to divide two of the enum values together?

Maybe I have expressed myself a bit badly. I don't want to apply arithmetics on two values of that enum. So nothing like `Red / Blue`.

But there are cases where enum values are used to calculate things.

```cs
enum CharacterType
{
    Player,
    Monster
}

int tableOffset = characterType * 32;
```

The calculations always take place in the domain of the integer, not the enum. But here the enum really act as an operand. And this is not that uncommon in games. You can also think of some character attribute that is calculated based on some enum.

**Reply by CyrusNajmabadi** [reply]
> Things like int color = ConsoleColor.Red; should be allowed. They are not really prone to errors

They definitely seem like it.  I have no idea how `.Red` is integer-esque.  What reasonable things would i do with that integer.  If i add '1' to it, can you tell me what that integer is?  If i dvide it by 2 what sense does that have?

What should i do with: `if (ReadInt() < colorAsInt)`?  Like, i'm honestly not seeing any world where that enum being an integer makes for sensible code.

 >For example when I have to add an offset to them, 

When are you adding an offset to ConsoleColor?  What does that mean?  What does it mean for StringSplitOptions?  I don't get it.

> increment them 

What does incrementing ConsoleColor mean?  What does it mean for UriComponents?  

> or assign them to the byte they are written to.

I don't know what that means, but serialization should not happen implicitly.  You should have to think about what it means (Esp. around endianness).  

> I am not sure if I get you right. I don't want everything to become ints. The enums stay as they are and can't be converted into other enums.

I wasn't talking about this.  I was talking about multiple enums mapping to the underlying type domain and then mixing them there.

**Reply by CyrusNajmabadi** [reply]
> Maybe I have expressed myself a bit badly. I don't want to apply arithmetics on two values of that enum. So nothing like Red / Blue.

I understand that.  But waht does it mean to do `ConsoleColor.Red / 32`?

**Reply by CyrusNajmabadi** [reply]
>But there are cases where enum values are used to calculate things.

this seems entirely domain specific to your domain and that *exact* enum.  That's something i would absolutely want to not be easy to do, but only happen through explicit code (ideally with comments) explaining what's going on and why it's ok for *exactly* this enum.  Indeed, i'd likely check that *only* that exact mathematical usage is allowed, and absolutely nothing else *at all ever*.

**Reply by FaustVX** [reply]
With the Roles & Extensions (#5497) proposal, will this be possible :
```C#
enum MyEnum { Value1, Value2 }
role CalculateEnum : MyEnum
{
  public static MyEnum operator /(MyEnum left, int scalar)
    => (MyEnum)((int)left / scalar);
}

var value = MyEnum.Value2 / 2; // value: Value1
```

**Reply by Pyrdacor** [reply]
> > Things like int color = ConsoleColor.Red; should be allowed. They are not really prone to errors
> 
> They definitely seem like it. I have no idea how `.Red` is integer-esque. What reasonable things would i do with that integer. If i add '1' to it, can you tell me what that integer is? If i dvide it by 2 what sense does that have?
> 
> What should i do with: `if (ReadInt() < colorAsInt)`? Like, i'm honestly not seeing any world where that enum being an integer makes for sensible code.
> 
> > For example when I have to add an offset to them,
> 
> When are you adding an offset to ConsoleColor? What does that mean? What does it mean for StringSplitOptions? I don't get it.
> 
> > increment them
> 
> What does incrementing ConsoleColor mean? What does it mean for UriComponents?
> 
> > or assign them to the byte they are written to.
> 
> I don't know what that means, but serialization should not happen implicitly. You should have to think about what it means (Esp. around endianness).
> 
> > I am not sure if I get you right. I don't want everything to become ints. The enums stay as they are and can't be converted into other enums.
> 
> I wasn't talking about this. I was talking about multiple enums mapping to the underlying type domain and then mixing them there.


It is not about that an enum value after incrementing or changing still makes sense in the domain of that enum. And that is not what I asked for. But it still makes sense in the domain of an integer. That's why I even convert it to an integer.

If I would not have a reason to convert it to an integer, I wouldn't do it.

Of course it does not make sense to have an integer of the value `Red` as it is. But dependent on that value some calculation might be necessary which follows a general arithmetic logic that can be easier expressed by using the underlying integral value as an operand. Or the enum is designed in such a way, that the values represent those integers which make sense as real values. A spell index enumeration expresses real indices but with the charme of implicit naming.

Enums in game data can be (and often are) designed in a way to make calculations, lookups or other things easier. Maybe this is niche but converting enums to ints would only performed where it makes some sense so I don't get the argument. Why would someone convert an enum to an int without a reason?

**Reply by Pyrdacor** [reply]
> With the Roles & Extensions (#5497) proposal, will this be possible :
> 
> ```cs
> enum MyEnum { Value1, Value2 }
> role CalculateEnum : MyEnum
> {
>   public static MyEnum operator /(MyEnum left, int scalar)
>     => (MyEnum)((int)left / scalar);
> }
> 
> var value = MyEnum.Value2 / 2; // value: Value1
> ```

This indeed looks promising. Thanks for the link. I'm fine with that.

**Reply by En3Tho** [reply]
So about this example:

```csharp
enum Error { NoError, AccessDenied, ... } 
int errorCode = Error.NoError;
// Some logic
Environment.Exit(errorCode);
```

It seems fine and neat and "logical" but the problem is it kinda ignores other space. For example, with extension everything feature you could just define a static extension member Exit that takes "Error" enum as parameter. With the way enums currently work now there is a clear separation between Exit(int) and Exit(error).

Also, I agree that it might lead to unexpected runtime errors due to making a typo. You basically lose a guarantee of explicitness and can introduce a bug where before you couldn't.

I get a feeling that you want some other feature instead, like a fixed set of concrete values, TS enums maybe? C# enums are kinda like that but at the same time they aren't.

**Reply by spydacarnage** [reply]
> But there are cases where enum values are used to calculate things.
> 
> ```cs
> enum CharacterType
> {
>     Player,
>     Monster
> }
> 
> int tableOffset = characterType * 32;
> ```
> 
> The calculations always take place in the domain of the integer, not the enum. But here the enum really act as an operand. And this is not that uncommon in games. You can also think of some character attribute that is calculated based on some enum.

This feels, to me, like it would be better suited as:
```cs
static class CharacterType
{
    const int Player = 0;
    const int Monster = 1;
}

int characterType = CharacterType.Player;
```

**Reply by CyrusNajmabadi** [reply]
> It is not about that an enum value after incrementing or changing still makes sense in the domain of that enum. And that is not what I asked for. But it still makes sense in the domain of an integer. That's why I even convert it to an integer.

> If I would not have a reason to convert it to an integer, I wouldn't do it.

But that's the thing, you're saying that nothing should prevent someone from converting an enum to an integer.  That's literally your request.  My point is that this restriction forces you to be explicit that you think this is ok.  You're saying we should always allow it because you'll only do it when it's ok.  But there's nothing you're providing that indicates how that remains true.  It basically is just "i won't ever screw this up, so let me do this".  whereas my position has been: "it looks like this would be a screwup in nearly all cases, and the only cases shown that wouldn't are incredibly niche, so preventing by default is absolutely the right hting".

This still seems to hold now.  We're not going to relax this if it's just for the incredibly niche cases that you yourself have said rarely hold, and especially not when you say that the only safeguard is that you simply won't screw this up.  That's precisely teh domain of wanting things to be explicit.  e.g.: very niche, more often than not the wrong thing, and requiring user vigilance to not screw things up.  Combined, all of this makes for a very poor choice to make something implicit.

**Reply by Pyrdacor** [reply]
Basically what I want is some kind of TS enums like @En3Tho said. The solution of @spydacarnage seems reasonable but has its own limitations. I would like to define a subset of values of some type X as a new type Y, where Y can be assigned to X as every value of Y is a value of X. You can think of it as a value-restricted type variation which allows the compiler to check valid value ranges and may avoid runtime value checks but still can be used in a context where the "source type" is usable. Maybe enums are not the right concept to use for that.

**Reply by Pyrdacor** [reply]
I guess it would be some extension of this:

```cs
class CharacterType : INumber<byte>, IMinMaxValue<byte>
{
    private readonly byte _value;

    private CharacterType(byte value)
    {
        _value = value;
    }

    public static readonly CharacterType Player = new CharacterType(0);
    public static readonly CharacterType Monster = new CharacterType(1);

    public static implicit operator byte(CharacterType characterType) => characterType._value;
    public static byte MaxValue => 1;
    public static byte MinValue => 0;

    // implement hundreds of other things for INumber<byte> here
}
```

But this is a lot of effort to allow such simple mechanic and it can't specify possible values with gaps. That's why I thought of enums. As you can basically define possible values in a convenient way there. But in the end it is all about value restriction while keeping the underlying type.

**Reply by spydacarnage** [reply]
If you have multiple types like this, you can use a source generator to construct all the boilerplate.  You could even hold a private enum inside your class (although honestly, I'd make it a struct so it acts more like a real `int`) to control what values the source generator creates, which could then include the gaps

**Reply by franchyd** [reply]
> > Still it is an integer as all values are integers. So I also want to be able to treat them as integers.
> 
> I'm curious why though. What's the value in treating it like an integer? Can it be a length? Can it be added to other values in that enum? What does it mean to divide two of the enum values together?

What is the value of being able to specify the underlying type of an enum then?  It is exclusively useful in scenarios where you will actually cast to/from that type.

**Reply by Pyrdacor** [reply]
> If you have multiple types like this, you can use a source generator to construct all the boilerplate. You could even hold a private enum inside your class (although honestly, I'd make it a struct so it acts more like a real `int`) to control what values the source generator creates, which could then include the gaps

Sure but it still seems like way too much effort for such simple thing. I mean I can also create my own kind of enum with a source generator as well. I could even implement my own integer class etc.

**Reply by HaloFour** [reply]
@DanFTRX 

> What is the value of being able to specify the underlying type of an enum then? It is exclusively useful in scenarios where you will actually cast to/from that type.

To explicitly specify the value space, and to allow proper translation in marshaling scenarios.  In the latter case the value may be treated as the raw integral type but that is not exposed in the C# code.

**Reply by spydacarnage** [reply]
> Sure but it still seems like way too much effort for such simple thing. I mean I can also create my own kind of enum with a source generator as well. I could even implement my own integer class etc.

Simple in appearance, but as it has already been determined that the language isn't going to change for such a minor edge-case, it doesn't seem like any more effort than you were hoping the language creators were going to employ.

And once you've written it, you'll be able to use that generator in every project you need it in, if it's _that_ important to you.

**Reply by HaloFour** [reply]
That particular flavor would likely need to be generated via source generators as it would likely run afoul of Oracle's [patent on class enums](https://patents.google.com/patent/US7263687B2/en) and couldn't be implemented as such in the language proper.

**Reply by Pyrdacor** [reply]
Very scary that companies have patents for such things.

**Reply by Pyrdacor** [reply]
> Simple in appearance, but as it has already been determined that the language isn't going to change for such a minor edge-case, 

I always wonder why I have so many minor edge cases in my projects and what the majority of programmers do. Most certainly I am doing things wrong. Or I hope that C# is something what it just not is. Which isn't meant in a bad way.

**Reply by CyrusNajmabadi** [reply]
You're not doing anything wrong. You're just doing niche things. Note that enums do work for your scenario, you're just asking for them guardrails (which can be trivially bypassed) be completely removed.

**Reply by Pyrdacor** [reply]
I wished there was a way in-between. Keeping it safe while still allow niche things. But I also have no clever idea. I can live with the current state. I had hope as you can already explicitly specify the underlying type. But I already supposed that it was just to limit the value range and for marshalling like HaloFour mentioned. There are a few things which are just relicts of the very early phase of C# so I hoped it is one of them. But your explanations seem reasonable so I'm fine with it.

**Reply by CyrusNajmabadi** [reply]
> I wished there was a way in-between. Keeping it safe while still allow niche things

There is.  Just cast to the underlying type.  it's normally just like 5 chars: `(int)`.  Incredibly trivial, easy to read and understand, and bridges the two worlds extremely well :)

**Reply by siegfriedpammer** [reply]
> That particular flavor would likely need to be generated via source generators as it would likely run afoul of Oracle's [patent on class enums](https://patents.google.com/patent/US7263687B2/en) and couldn't be implemented as such in the language proper.

We only have to wait another 17 months for the patent to expire and then class enums could be implemented, right?

**Reply by HaloFour** [reply]
> We only have to wait another 17 months for the patent to expire and then class enums could be implemented, right?

Theoretically, although I personally believe that discriminated unions could cover pretty much all of the use cases where such enums might be useful.

**Reply by Pyrdacor** [reply]
I never liked them in Java and I wouldn't vote for them. I rather would like something which can just inherit the primitive types and restrict or extend them in some way.
